File: CPP/Graphs/Flows/Dinic.cpp
// Dinic
//
template <typename T>
class Dinic {
  public:
	FlowGraph<T>& g;

	vector<int> ptr;
	vector<int> level;
	vector<int>q;

	Dinic(FlowGraph<T>& _g) : g(_g) {
		ptr.resize(g.n);
		level.resize(g.n);
		q.resize(g.n);
	}

	bool bfs() {
		fill(level.begin(), level.end(), -1);
		q[0] = g.st;
		level[g.st] = 0;
		int beg = 0, end = 1;
		while (beg < end) {
			int u = q[beg++];
			for (int id : g.g[u]) {
				const auto& e = g.edges[id];
				const auto& back = g.edges[id ^ 1];
				if (e.c - e.f > g.eps && level[e.to] == -1) {
					level[e.to] = level[u] + 1;
					if (e.to == g.fin) {
						return true;
					}
					q[end++] = e.to;
				}
			}
		}
		return false;
	}

	T dfs(int v, T w) {
		if (v == g.fin) {
			return w;
		}
		int& j = ptr[v];
		while (j < int(g.g[v].size())) {
			int id = g.g[v][j];
			const auto& e = g.edges[id];
			if (e.c - e.f > g.eps && level[e.to] == level[v] + 1) {
				T t = dfs(e.to, min(e.c - e.f, w));
				if (t > g.eps) {
					g.edges[id].f += t;
					g.edges[id ^ 1].f -= t;
					return t;
				}
			}
			j++;
		}
		return 0;
	}

	T max_flow() {
		while (bfs()) {
			fill(ptr.begin(), ptr.end(), 0);
			T big_add = 0;
			while (true) {
				T add = dfs(g.st, numeric_limits<T>::max());
				if (add <= g.eps) {
					break;
				}
				big_add += add;
			}
			if (big_add <= g.eps) {
				break;
			}
			g.flow += big_add;
		}
		return g.flow;
	}

	vector<bool> min_cut() {
		max_flow();
		vector<bool> ret(g.n);
		for (int i = 0; i < g.n; i++) {
			ret[i] = (level[i] != -1);
		}
		return ret;
	}
};

File: Kotlin/Graphs/Dinic.kt
class Dinic {
	var g:FlowGraph
	var ptr:IntArray
	var level:IntArray
	var q:IntArray

	constructor(g_:FlowGraph) {
		g = g_
		ptr = IntArray(g.n){0}
		level = IntArray(g.n){0}
		q = IntArray(g.n){0}
	}

	fun bfs():Boolean {
		level = IntArray(g.n){-1}
		q[0] = g.st
		level[g.st] = 0
		var beg = 0
		var end = 1
		while (beg < end) {
			val u = q[beg++]
			for (id in g.g[u]) {
				val e = g.edges[id]
				val back = g.edges[id xor 1]
				if (e.c - e.f > g.eps && level[e.to] == -1) {
					level[e.to] = level[u] + 1;
					if (e.to == g.fin) {
						return true
					}
					q[end++] = e.to
				}
			}
		}
		return false
	}

	fun dfs(v:Int, w:Long):Long {
		if (v == g.fin) {
			return w
		}
		while (ptr[v] < g.g[v].size) {
			val id = g.g[v][ptr[v]]
			val e = g.edges[id]
			if (e.c - e.f > g.eps && level[e.to] == level[v] + 1) {
				val t = dfs(e.to, Math.min(e.c - e.f, w))
				if (t > g.eps) {
					g.edges[id].f += t
					g.edges[id xor 1].f -= t
					return t
				}
			}
			ptr[v]++
		}
		return 0L
	}

	fun max_flow():Long {
		while (bfs()) {
			ptr = IntArray(g.n){0}
			var big_add = 0L
			while (true) {
				val add = dfs(g.st, Long.MAX_VALUE)
				if (add <= g.eps) {
					break
				}
				big_add += add
			}
			if (big_add <= g.eps) {
				break
			}
			g.flow += big_add
		}
		return g.flow
	}

	fun min_cut():BooleanArray {
		max_flow()
		val ret = BooleanArray(g.n){false}
		for (i in 0 until g.n) {
			ret[i] = (level[i] != -1)
		}
		return ret
	}
}